Performance optimization

textGamex opened the pull request.

textGamex committed:
perf: Improve file extension check for rule files in loadBaseConfig function

textGamex committed:
perf: Remove unused debugChecks function for improved performance

textGamex committed:
perf: Remove unused mutable variables for improved performance

textGamex committed:
refactor: Remove unused recursive functions for improved performance

textGamex committed:
perf: Replace mutable dictionary with FrozenDictionary for improved performance

textGamex commented:
@tboby Can we add support for the .NET 9 framework while retaining .NET 8? Some APIs, such as `system.threading.lock`, are only available in .NET 9.

tboby commented:
Multi-targeting and using define blocks is possible, but it's messy. If there's a useful feature we can upgrade to net9.0: net10 is only 3 months away anyway.

textGamex commented:
Also, just to confirm, LSP project can be upgraded to .NET 9, right?

tboby commented:
Yup

textGamex committed:
perf: Update string quotation check to use single quotes for improved performance

textGamex committed:
perf: Optimize alias extraction by using Seq for improved performance

textGamex committed:
refactor: Remove unused type alias for improved performance

textGamex committed:
fix error

textGamex committed:
perf: remove unneeded ToFrozenDictionary()

textGamex committed:
perf: Simplify struct definitions for StringTokens and StringMetadata and add IsReadOnly

textGamex committed:
perf: Replace LowerCaseStringSet with IgnoreCaseStringSet for improved case handling

textGamex committed:
refactor: clean up code and remove unused method params

textGamex committed:
perf: use single quotes for character literals in FieldValidators.fs

textGamex committed:
perf: optimize locking mechanism in FileIndexTable and StringTokens intern method

textGamex committed:
fix error

tboby commented:
@textGamex Which IDE are you using for F#/C#? Just want to make sure I set up the auto-formatting for it too.

textGamex commented:
> Which IDE are you using for F#/C#? Just want to make sure I set up the auto-formatting for it too.

I use Rider.

tboby commented:
Merged the formatting: sorry, conflicts!

With rider it *should* just automatically use `fantomas` if you trigger format. I have my rider configured to format all files on commit, which seems to work well.

One of the files doesn't format: we found a bug in the formatter :)

textGamex committed:
Merge branch 'master' into perf

tboby commented:
Why `NET5_0_OR_GREATER` in the `CWTools` project? The project is `net8.0` and soon `net9.0`, so no need to use defines :)

textGamex commented:
> Why NET5_0_OR_GREATER in the CWTools project? The project is net8.0 and soon net9.0, so no need to use defines :)

This is a legacy issue from when the project was still supporting .NET Standard 2.0. I will fix this issue.

textGamex committed:
fix error

textGamex committed:
refactor: remove NET5_0_OR_GREATER

textGamex committed:
refactor: simplify access to Map values in RulesManager.fs

textGamex committed:
refactor: update ModifierCategory and Scope types for improved property access

textGamex commented:
@tboby Can we always add the `sealed` modifier to a class, or should we consider the potential users of the CWTools library?

tboby commented:
> @tboby Recently, I've been playing RimWorld and got quite engrossed in it, which caused the project development to be put on hold for a few days.
> 
> Can we introduce GitHub's dependency bot to automatically update dependencies?

Hah, don't worry, I have [plenty ](https://github.com/tboby/wyoming-onnx-asr )of [things ](https://github.com/tboby/fantomas/tree/issue3174-simple) that distract me too :)

Unfortunately dependabot doesn't support paket, otherwise yes. It should be set up for the npm packages in the other repo.

>  @tboby Can we always add the sealed modifier to a class, or should we consider the potential users of the CWTools library?

Yup, that should be fine. We can always unseal if it becomes a problem, and/or they can fork and build their own while we consider :)

textGamex committed:
refactor: remove unused validation files from project

textGamex committed:
perf： add sealed keyword

textGamex committed:
refactor: clean up Position.fs by removing unused modules and simplifying path checks

textGamex committed:
perf: add sealed attribute to multiple types for improved performance

textGamex committed:
perf: update staticModifierCategory to use node.Leaves

textGamex committed:
perf: replace union type with enum for EffectType and TitleType

tboby commented:
I've been trying to do some regression testing (checking that the same number of errors are produced with the new `cwtools` library version), and have discovered that the localisation parser change is causing problems.
https://github.com/cwtools/cwtools/commit/d54e9faf198a5c73a209ef2bbbcb99bb1326aedc#diff-b4db0dbb70db1d558973f6767993a4a9d61ff120001d55b2063a7362f4631cf5R32

I haven't been able to work out why yet, but the change to `desc` is basically failing to parse all localisation in stellaris.

textGamex commented:
> I haven't been able to work out why yet, but the change to desc is basically failing to parse all localisation in stellaris.

Is this related to the issue mentioned in #85?

textGamex committed:
perf: use array replace list in rule parser

textGamex committed:
Merge branch 'master' into perf

textGamex committed:
perf: Optimize string split performance

textGamex committed:
perf: Improve case-insensitive check for definition.csv file

textGamex committed:
perf: Optimize string set creation for array input

textGamex committed:
perf: Refactor data structures to use arrays instead of lists for improved performance

textGamex committed:
perf: use array.Length replace List.Length

textGamex committed:
perf: replace HashSet with FrozenSet for case-insensitive string set

textGamex committed:
perf: use KTrie replace VDS.Common.Tries

textGamex commented:
@tboby I have introduced a new third-party library 'KTrie' to replace the previously used 'VDS.Common.Tries' for performance optimization, but I am not sure how to correctly add the dependency. Could you please help me add the KTrie dependency to my project?

textGamex committed:
perf: seal CharacterComparer type

textGamex committed:
perf: use span to avoid memory allocations when trimming strings.

tboby committed:
infra: move KTrie to paket, remove VDS

textGamex commented:
> FYI: I'm getting
> Runtime error: One or more errors occurred. (The value cannot be an empty string. (Parameter 'word'))
> in my test project (Stellaris + Gigastructures) probably from trie.Contains.

I will fix this bug (by adding additional checks for empty strings). Additionally, are there any unique features of Paket that are essential for the project? Using NuGet instead might lower the entry barrier for the project.

tboby commented:
> Additionally, are there any unique features of Paket that are essential for the project? Using NuGet instead might lower the entry barrier for the project.

Nuget doesn't support dependencies on commits, so can't use it for cwtools-vscode without setting up an excessive versioning system for cwtools.

For cwtools though, there is only on git commit dependency which I could just vender, so more plausible. Central package management works fine in IDEs, but I don't know how well lock files are supported. I'll have to experiment with how IDEs cope with lock files.

textGamex committed:
fix: throw exception when param is empty string

textGamex committed:
perf: replace Collections.Set with FrozenSet for improved performance

textGamex committed:
perf: use ZString for improved file path validation

textGamex committed:
perf: optimize fLeaf,fLeafValue,fNode of getTypesInEntity

textGamex committed:
perf: simplify ampersand character handling in Scopes.fs

textGamex committed:
perf: introduce SplitFirst extension method for improved character splitting

textGamex committed:
perf: Overwrite use enum type

textGamex commented:
Perhaps we can update the FParsec package to version 2.0.0 beta 2, in which I haven't encountered any issues.

textGamex committed:
deps: update FParsec and packages of test project

textGamex committed:
Merge branch 'master' into perf

textGamex committed:
perf: update Microsoft.Testing.Platform and MSBuild to version 1.8.2 in paket.lock

textGamex commented:
@tboby "After merging the changes from the main branch, the perf project no longer works with the original parameters. Are there any breaking changes?
Error message: Argument parsing error: no valid subcommand has been specified.
Parameters: `-hoi4 -mod-path "C:\Users\QWQ\Desktop\Mod" -g "D:\SteamLibrary\steamapps\common\Hearts of Iron IV" -p "D:\Code\Project\GitProject\cwtools-vscode\release\.cwtools\hoi4.cwb" -c "D:\Code\Project\GitProject\cwtools-vscode\release\.cwtools\hoi4\Config"`

textGamex commented:
> You just need to add full to the... start or end?

No matter added at the start or the end, it prompts: "Argument parsing error: no valid subcommand has been specified."

tboby commented:
> > You just need to add full to the... start or end?
> 
> No matter added at the start or the end, it prompts: "Argument parsing error: no valid subcommand has been specified."

```
    | [<CliPrefix(CliPrefix.None)>] Full
    | [<CliPrefix(CliPrefix.None); CustomCommandLine("localisation")>] Localisation of ParseResults<Nothing>
```

You could try modifying it to match the localisation subcommand, or just temporarily undo the changes to that project and I'll fix it again later :)

textGamex committed:
perf: optimize icon field checks using StringBuilder for improved performance

textGamex committed:
perf: replace quoteCharArray with single quoteChar for improved clarity

textGamex committed:
perf(Interning): use GetOrAdd replace raw implementation

textGamex committed:
style: simplify array access syntax

textGamex committed:
perf(RulesParser): simplify substring access syntax for improved perf

textGamex committed:
perf(ResourceManager): optimize glob pattern matching using ConcurrentDictionary for improved performance

textGamex committed:
perf(FieldValidatorsHelper): optimize string comparison using AsSpan for improved performance

textGamex committed:
perf(FileManager): FilesScope change to enum

textGamex committed:
perf(FileManager): add FileManagerHelper for path conversion and optimization

textGamex committed:
perf(FileManager): optimize file enumeration by using Directory.EnumerateFiles for improved performance

textGamex committed:
perf(FileManager): optimize directory enumeration by replacing recursive folder retrieval with Directory.EnumerateDirectories

textGamex committed:
perf(Utilities): remove unused stopwatch for duration measurement

textGamex committed:
Refactor changeScope usage across multiple modules

- Updated the definition of changeScope in VIC2Scopes.fs and VIC3Scopes.fs to use a unified ChangeScope type.
- Refactored calls to changeScope in CompletionService.fs, FieldValidators.fs, InfoService.fs, and RuleValidationService.fs to use the new Invoke method.
- Improved readability by applying consistent formatting and removing unnecessary comments.
- Introduced parseIntSpan and parseDecimalSpan functions in Utilities.fs for better handling of ReadOnlySpan<char>.

textGamex committed:
perf(RuleValidationService): remove unused mutable errorList

textGamex committed:
perf(Scopes): simplify key parsing by removing unused variables

textGamex committed:
perf(Scopes): optimize struct usage in scope functions for better performance

textGamex committed:
perf(FieldValidators): simplify value type handling by removing unused parameters

textGamex committed:
perf(CompletionService): simplify string splitting by removing unnecessary array syntax

textGamex committed:
perf(CommonValidation): optimize BoolState usage by replacing direct values with enum members

textGamex committed:
perf(ResourceManager): Reduce memory allocation when getting run time

textGamex committed:
style: use _ replace __ in member

textGamex committed:
perf(ResourceManager): combine filters to improve entity validation efficiency

textGamex committed:
perf(Validation): optimize EntitySet memory usage

textGamex committed:
perf(ResourceManager): seal ResourceManager class and remove NoInlining attribute from InternIdentifierToken method

textGamex committed:
perf(STLProcess, STLEvents): optimize string handling for event target extraction

textGamex committed:
fix match bug

textGamex committed:
perf(ModifierCategoryManager): optimize dictionary handling with case-insensitive key comparison

textGamex committed:
perf: optimize array handling by replacing List with Array for improved performance

textGamex committed:
perf: improve position comparison by using Equals method for accuracy

textGamex committed:
perf: optimize collection handling by replacing Children with Nodes for improved performance

textGamex committed:
perf: optimize script folder handling by replacing lists with arrays for improved performance

textGamex committed:
perf: remove debug log statement from CreateGame method

textGamex committed:
perf: optimize ignoreGlobList handling by replacing lists with arrays for improved performance

textGamex committed:
perf: optimize language handling by replacing lists with arrays for improved performance

textGamex committed:
perf: optimize variable collection by replacing Seq.collect with List.collect for improved performance

textGamex committed:
perf: optimize CheckPathDir method by reducing unnecessary array allocation and improving path handling

tboby commented:
> @tboby I'd like to enable trimming for project, but I'm not sure how to adjust the build process to use the publish command instead of build. I need your help with this.
> 
> Based on my testing, enabling `partial` trimming can reduce the assembly size from 105 MB to 63.5 MB. (The `full` trimming mode causes the application to fail, likely requiring some code adjustments.)
> 
> You just need to add these two lines to the .fsproj file of the Main project to enable the trimming feature.
> 
> ```
> <PublishTrimmed>true</PublishTrimmed>
> <TrimMode>partial</TrimMode>
> ```

I've enabled this on the extension :)

Sorry for the delays, I was last minute planning for, then I was on holiday. I've released a new version of cwtools-vscode, so the chinese loc + the previous optimizations MR has been published.

tboby commented:
I can also make the package 1/3 the size if I use
https://code.visualstudio.com/api/working-with-extensions/publishing-extension#platformspecific-extensions

On Mon, 15 Sept 2025, 06:25 textGamex, ***@***.***> wrote:

> *textGamex* left a comment (cwtools/cwtools#86)
> <https://github.com/cwtools/cwtools/pull/86#issuecomment-3290536472>
>
> Oh, I think I was mistaken. I see that the file size has indeed decreased.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cwtools/cwtools/pull/86#issuecomment-3290536472>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADNHH3CI2AAYSUTFI5RVZL3SZEUNAVCNFSM6AAAAACCNPJ262VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTEOJQGUZTMNBXGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

textGamex committed:
fix #47 issue

textGamex committed:
Merge branch 'refs/heads/master' into perf

# Conflicts:
#	CWTools/Common/HOI4Constants.fs
#	CWTools/Serialization/Serializer.fs

textGamex committed:
fix merge error

textGamex committed:
perf: use use keyword replace using()

textGamex committed:
perf: improve path handling and optimize directory checks in addDLCs function

textGamex committed:
perf: remove unused rulesLocalisationErrors property from LocalisationManager

textGamex committed:
perf: replace InsensitiveStringComparer with StringComparer

textGamex committed:
fix: missing file name in CW113 Error message

textGamex committed:
perf: replace List.fold with Seq.fold for improved performance in foldNode7 function

textGamex committed:
perf: optimize child counting and clause retrieval in CompletionService and RuleValidationService

textGamex committed:
perf: optimize clause processing in RuleValidationService by removing unnecessary List conversion

tboby commented:
> @tboby Are there any known bugs in this PR? If not, could we merge it into the main branch? After all, resolving merge conflicts is a hassle, especially for such a large PR.

Sorry, you're right. I've done the updating for cwtools-vscode: just trying to pin down a change in scope validation in localisation strings. I can't tell if it's a fix or a bug :)

textGamex committed:
perf: streamline path handling in CompletionService and RuleValidationService by removing redundant variables

textGamex committed:
refactor: Only 0 is allowed in CK2DNA rules, not all numbers

textGamex committed:
refactor: eliminate unnecessary List conversion

textGamex committed:
perf: optimize value handling in CompletionService and FieldValidators by using arrays instead of lists

tboby commented:
I'm also seeing higher peak memory usage, but when I profile the steady-state memory usage is the same or lower. This would be fine except I can't get the runtime to release memory well, so it's sitting there at 7GB when it's happy with 4GB and only has ~2GB actually allocated.
<img width="2880" height="1760" alt="image" src="https://github.com/user-attachments/assets/ebc31d78-7f1f-4513-8917-79c07d83bda2" />

I have a GC.Collect which is supposed to solve this, but it isn't enough apparently.

textGamex commented:
I don’t understand why the runtime requested 7GB but is only using 2GB.

textGamex committed:
perf(localisation): optimize memory usage of YAMLLocalisationService

textGamex commented:
> I'm struggling the open the latest memory profile (it's too big on stellaris without mods!), but it looks like there's 1.3GB of tuple ruletype x options, which seems a bit much.

Yes, I also noticed this phenomenon while doing performance optimization and attempted to optimize it. However, due to the system being highly coupled and my limited familiarity with the project, I ultimately failed.

textGamex committed:
perf: inline functions for key retrieval in FieldValidators to improve performance

textGamex committed:
perf: optimize string replacement in FileManager and Game modules for improved readability

textGamex committed:
perf: streamline file parsing and resource handling in Game module for improved efficiency

textGamex committed:
perf: refactor localisation validation functions for improved clarity and performance

textGamex committed:
perf: enhance FileManager for improved directory classification and path normalization

textGamex commented:
@tboby Oh, by the way, if you’re going to optimize this(tuple ruletype x options), could you make the changes based on this PR? That way, I won’t have to resolve any merge conflicts.

textGamex committed:
delete unused file

textGamex committed:
fix error

textGamex committed:
perf: optimize comment handling and improve constant definitions in RulesParser

textGamex committed:
perf: remove unused character sets in SharedParsers for improved clarity

textGamex committed:
perf: simplify LocalisationValidation functions by removing unnecessary parameters

textGamex committed:
perf: replace Array.choose with Seq.choose for improved performance in InfoService

textGamex committed:
Revert "perf: replace Array.choose with Seq.choose for improved performance in InfoService"

This reverts commit 79be1016a78c6fe03b128c89423688e8d4c9a7dc.

textGamex committed:
perf: remove extra ToArray()

textGamex committed:
perf: use struct tuple for optimize performance in LocalisationAPIs

textGamex committed:
refactor: optimize  createEmbeddedSettings method of all games

textGamex commented:
@tboby Is the file `CWToolsCSTests/Program.cs` still in use? If not, I’d like to clean it up to reduce the refactoring workload.

textGamex committed:
perf: simplify array access in character extraction logic

textGamex committed:
perf: replace List with Array for improved performance in scope and modifier handling

textGamex committed:
perf: use case-insensitive string comparison for dictionary lookups in ScopeManager

textGamex committed:
perf: use struct tuples and improve equality checks for Scope comparisons

textGamex committed:
perf: remove unused ScopeWrapper type to streamline Constants.fs

textGamex committed:
refactor: refactor language arrays to use Enum.GetValues for improved maintainability

textGamex committed:
fix: add support for EU5 and VIC3 languages in serializer

textGamex committed:
refactor: remove unused code

textGamex committed:
deps: update Microsoft.Testing.Platform and MSBuild to version 1.8.4 for improved compatibility

textGamex committed:
perf: optimize Lookup.typeDefInfoForValidation

textGamex committed:
perf: replace Bool true/false with YesBool/NoBool

textGamex committed:
Revert "perf: replace Bool true/false with YesBool/NoBool"

This reverts commit 74797005889714fdefb1f83e6e4a215dbce82f72.

textGamex committed:
perf: rename functions for clarity in Lookups.fs

textGamex committed:
perf: remove unused definedScriptVariables from Lookups.fs

textGamex committed:
perf: refactor modifier and feature settings loading in game files

textGamex committed:
perf: simplify Key type

textGamex committed:
perf: remove unused NodeTypeMap and updateContext functions from Process.fs

textGamex committed:
perf: change the type of Lookup.coreModifiers from list to array.

textGamex committed:
perf: remove unused Modifier type from Constants.fs

textGamex committed:
perf: change StaticModifiers from list to array

textGamex committed:
perf: change scriptedLoc from list to array

textGamex committed:
perf: clean up CK2Localisation.fs by removing commented-out code

textGamex committed:
perf: streamline localisation handling by introducing GetEntries method and updating related functions

textGamex committed:
perf: optimize localisation entry processing by replacing List.collect with Seq.collect

textGamex committed:
perf: optimize ResourceManager by replacing Map with Dictionary for fileMap

textGamex committed:
perf: optimize ResourceManager by replacing Map with Dictionary for entitiesMap and improving data access patterns

textGamex committed:
perf: optimize ResourceManager by replacing List operations with Array for improved performance

textGamex committed:
perf: optimize game settings by replacing List with Array for rootDirectories to enhance performance

textGamex committed:
perf: optimize province file processing in Hoi4

tboby commented:
> @tboby Is the file `CWToolsCSTests/Program.cs` still in use? If not, I’d like to clean it up to reduce the refactoring workload.

No, that can go.

> defaultScopes and defaultModifiers

Yup, they can go, as long as that game actually has them in rules xD

textGamex commented:
> My guess is that one change made things a lot worse and is bigger than all the small improvements.

My feeling is that this issue is caused by the conversions between lists and arrays. Previously, most of the code used lists. Now, I have replaced some lists with arrays and resizeArray, and performed conversions at the interface points. If we convert all lists to arrays and resizeArray in the future, it might improve this situation.

textGamex committed:
Merge branch 'master' into perf

textGamex committed:
perf: remove unused variable 'useRules' to streamline code

textGamex committed:
Merge remote-tracking branch 'origin/perf' into perf

textGamex committed:
refactor: delete unused file

textGamex committed:
perf: remove redundant variable assignment to simplify code

textGamex committed:
perf: optimize entity lookup in InfoAtPos

textGamex committed:
perf: replace List with Seq for improved performance in entity processing

textGamex committed:
perf: convert Seq to array for improved performance in getEnumsFromComplexEnums

textGamex committed:
feat: add ByteSize package for improved memory reporting in performance tests

textGamex commented:
@tboby Can we integrate telemetry to see which games are used more frequently by users, so that we can determine the optimization priorities?

textGamex committed:
refactor: remove unused function

textGamex committed:
perf: convert List to Array for improved performance in localisation handling

textGamex committed:
perf: remove unused variables and functions for improved code clarity

textGamex committed:
perf: optimize file extension checks for improved performance in localisation handling

textGamex committed:
deps: update package

textGamex committed:
perf: optimize data structures by converting lists to arrays for improved performance in localisation handling

textGamex committed:
perf: LocKeySet replace to HashSet

textGamex committed:
perf: streamline localisation key processing

textGamex committed:
perf: optimize localisation processing by using Map.ofArray and accessing values directly
